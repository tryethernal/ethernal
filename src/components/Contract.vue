<template>
    <v-container fluid>
        <v-alert v-if="notAContract" dense text type="warning">There doesn't seem to be a contract at this address</v-alert>
        <v-row class="mb-1">
            <v-col cols="12" lg="5">
                <v-card outlined style="height: 100%">
                    <v-skeleton-loader v-if="loadingContract" type="list-item"></v-skeleton-loader>
                    <template v-if="contract.name">
                        <v-card-title>
                            {{ contract.name }}
                            <template v-if="isUserAdmin">
                                <v-spacer></v-spacer>
                                <Remove-Contract-Confirmation-Modal ref="removeContractConfirmationModal" />
                                <v-btn x-small icon @click="openRemoveContractConfirmationModal()">
                                    <v-icon color="red">mdi-delete</v-icon>
                                </v-btn>
                            </template>
                        </v-card-title>
                    </template>
                    <v-card-subtitle style="display: flex;" class="justify-space-between">
                        <v-skeleton-loader v-if="loadingContract" type="chip"></v-skeleton-loader>
                        <v-chip v-else v-for="(pattern, idx) in contract.patterns" :key="idx" x-small class="success mr-2">
                            {{ formatContractPattern(pattern) }}
                        </v-chip>
                        <template v-if="isUserAdmin && !contract.name">
                            <v-spacer></v-spacer>
                            <Remove-Contract-Confirmation-Modal ref="removeContractConfirmationModal" />
                            <v-btn x-small icon @click="openRemoveContractConfirmationModal()">
                                <v-icon color="red">mdi-delete</v-icon>
                            </v-btn>
                        </template>
                    </v-card-subtitle>
                    <v-card-text>
                        <v-row>
                            <v-col v-if="isErc20 || isErc721" cols="6">
                                <small>Token Name</small><br>
                                <v-skeleton-loader v-if="loadingContract" type="list-item"></v-skeleton-loader>
                                <span v-else-if="contract.tokenName" class="ml-2">
                                    <Hash-Link :type="tokenType" :hash="contract.address" :withName="true" :withTokenName="true" />
                                </span>
                                <span v-else>
                                    N/A
                                </span>
                            </v-col>

                            <v-col v-if="isErc20 || isErc721" cols="6">
                                <small>Token Symbol</small><br>
                                <v-skeleton-loader v-if="loadingContract" type="list-item"></v-skeleton-loader>
                                <span v-else class="text-h6 ml-2">{{ contract.tokenSymbol || 'N/A' }}</span>
                            </v-col>


                            <v-col cols="6">
                                <small>Address</small><br>
                                <v-skeleton-loader v-if="loadingContract" type="list-item"></v-skeleton-loader>
                                <span v-else class="ml-2">
                                    <Hash-Link :type="'address'" :hash="address" />
                                </span>
                            </v-col>

                            <v-col cols="6">
                                <small>Contract Creation</small><br>
                                <v-skeleton-loader v-if="loadingContract" type="list-item"></v-skeleton-loader>
                                <span v-else-if="contract.creationTransaction && contract.creationTransaction.hash" class="ml-2">
                                    <Hash-Link :type="'transaction'" :hash="contract.creationTransaction.hash" />
                                </span>
                                <span v-else class="ml-2">N/A</span>
                            </v-col>
                        </v-row>
                    </v-card-text>
                </v-card>
            </v-col>

            <v-col cols="12" lg="7">
                <v-row>
                    <v-col cols="12">
                        <v-card outlined>
                            <v-card-subtitle>Balance</v-card-subtitle>
                            <v-skeleton-loader v-if="loadingStats" type="list-item"></v-skeleton-loader>
                            <v-card-text v-else class="text-h4" align="center">
                                {{ balance | fromWei('ether', chain.token) }}
                            </v-card-text>
                        </v-card>
                    </v-col>
                </v-row>

                <v-row>
                    <v-col cols="12" lg="6">
                        <v-card outlined style="height: 100%;">
                            <v-card-subtitle>Transactions</v-card-subtitle>
                            <v-skeleton-loader v-if="loadingStats" type="list-item"></v-skeleton-loader>
                            <v-card-text v-else class="text-h4" align="center">
                                <template v-if="!contract">
                                    {{ sentTransactionCount }}<v-icon>mdi-arrow-up-thin</v-icon>
                                    <v-divider vertical class="mx-4"></v-divider>
                                </template>
                                {{ receivedTransactionCount }}<v-icon v-if="!contract">mdi-arrow-down-thin</v-icon>
                            </v-card-text>
                        </v-card>
                    </v-col>

                    <v-col cols="12" lg="6">
                        <v-card outlined style="height: 100%;">
                            <v-card-subtitle>ERC-20 Transfers</v-card-subtitle>
                            <v-skeleton-loader v-if="loadingStats" type="list-item"></v-skeleton-loader>
                            <v-card-text v-else class="text-h4" align="center">
                                {{ sentErc20TokenTransferCount }}<v-icon>mdi-arrow-up-thin</v-icon><v-divider vertical class="mx-4"></v-divider> {{ receivedErc20TokenTransferCount }}<v-icon>mdi-arrow-down-thin</v-icon>
                            </v-card-text>
                        </v-card>
                    </v-col>
                </v-row>
            </v-col>
        </v-row>

        <v-tabs v-model="tab">
            <v-tab id="transactionsTab" href="#transactions">Transactions</v-tab>
            <v-tab id="interactionsTab" href="#interactions">Read / Write</v-tab>
            <v-tab id="logsTab" href="#logs">Logs</v-tab>
            <v-tab id="codeTab" href="#code">Code</v-tab>
            <v-tab id="storageTab" href="#storage">Storage</v-tab>
        </v-tabs>

        <v-tabs-items :value="tab">
            <v-tab-item value="transactions">
                <Address-Transactions-List :address="address" />
            </v-tab-item>

            <v-tab-item value="interactions">
                <Contract-Interaction :address="address" />
            </v-tab-item>

            <v-tab-item value="logs">
                <Contract-Logs :address="address" />
            </v-tab-item>

            <v-tab-item value="storage">
                <Contract-Storage :address="address" />
            </v-tab-item>

            <v-tab-item value="code">
                <Contract-Code v-if="contract" :contract="contract" />
            </v-tab-item>
        </v-tabs-items>
    </v-container>
</template>

<script>
const moment = require('moment');
const ethers = require('ethers');
import { mapGetters } from 'vuex';

const { formatNumber, formatContractPattern } = require('../lib/utils');

import AddressTransactionsList from './AddressTransactionsList';
import ContractInteraction from './ContractInteraction';
import ContractStorage from './ContractStorage';
import ContractLogs from './ContractLogs';
import ContractCode from './ContractCode';
import HashLink from './HashLink';
import FromWei from '../filters/FromWei';
import RemoveContractConfirmationModal from './RemoveContractConfirmationModal';

export default {
    name: 'Contract',
    props: ['address'],
    components: {
        AddressTransactionsList,
        HashLink,
        ContractInteraction,
        ContractLogs,
        ContractStorage,
        ContractCode,
        RemoveContractConfirmationModal
    },
    filters: {
        FromWei
    },
    data: () => ({
        balance: 0,
        loadingContract: true,
        loadingStats: true,
        loadingBalance: false,
        contract: {
            patterns: []
        },
        receivedTransactionCount: null,
        sentErc20TokenTransferCount: null,
        receivedErc20TokenTransferCount: null,
        metamaskData: {},
        connectedAccountBalance: null,
        notAContract: false
    }),
    mounted() {
        this.server.getAccountBalance(this.address).then(balance => this.balance = ethers.BigNumber.from(balance).toString());
    },
    methods: {
        moment: moment,
        formatNumber: formatNumber,
        formatContractPattern: formatContractPattern,
        openRemoveContractConfirmationModal() {
            this.$refs.removeContractConfirmationModal
                .open({ address: this.address, workspace: this.currentWorkspace.name });
        }
    },
    watch: {
        address: {
            immediate: true,
            handler(address) {
                this.server.getContract(address)
                    .then(({ data }) => {
                        if (data)
                            this.contract = data
                        else
                            this.notAContract = true;
                    })
                    .finally(() => this.loadingContract = false);

                this.server.getAddressStats(address)
                    .then(({ data }) => {
                        this.receivedTransactionCount = data.receivedTransactionCount;
                        this.sentErc20TokenTransferCount = data.sentErc20TokenTransferCount;
                        this.receivedErc20TokenTransferCount = data.receivedErc20TokenTransferCount;
                    })
                    .finally(() => this.loadingStats = false);
            }
        }
    },
    computed: {
        ...mapGetters([
            'currentWorkspace',
            'chain',
            'isUserAdmin'
        ]),
        isErc20() {
            return this.contract.patterns.indexOf('erc20') > -1;
        },
        isErc721() {
            return this.contract.patterns.indexOf('erc721') > -1;
        },
        tokenType() {
            if (this.isErc20) return 'token';
            if (this.isErc721) return 'nft';
            else return 'contract';
        },
        tab: {
            set(tab) {
                this.$router.replace({ query: { ...this.$route.query, tab } }).catch(()=>{});
            },
            get() {
                return this.$route.query.tab;
            }
        }
    }
}
</script>
