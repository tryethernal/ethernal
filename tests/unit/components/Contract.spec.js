import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

jest.mock('@metamask/detect-provider');
import Contract from '@/components/Contract.vue';

let helper;
const stubs = [
    'Import-Artifact-Modal',
    'Remove-Contract-Confirmation-Modal',
    'Hash-Link',
    'Contract-Interaction',
    'Contract-Logs',
    'Contract-Storage',
    'Contract-Code',
    'Address-Transactions-List'
];

describe('Contract.vue', () => {

    beforeEach(() => {
        jest.clearAllMocks();
        helper = new MockHelper();
        jest.spyOn(helper.mocks.server, 'getAccountBalance')
            .mockResolvedValue('10000');
        jest.spyOn(helper.mocks.server, 'getAddressStats')
            .mockResolvedValue({ data: {
                sentTransactionCount: 1,
                receivedTransactionCount: 2,
                sentErc20TokenTransferCount: 3,
                receivedErc20TokenTransferCount: 4
            }});
    });

    it('Should display a warning if not a contract', async () => {
        jest.spyOn(helper.mocks.server, 'getContract')
            .mockResolvedValueOnce({ data: null });
        const wrapper = helper.mountFn(Contract, {
            propsData: {
                hash: '0x123'
            },
            stubs: stubs
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display only address & creation if not a token', (done) => {
        jest.spyOn(helper.mocks.server, 'getContract')
            .mockResolvedValueOnce({ data: {
                name: 'Contract',
                patterns: [],
                address: '0x123',
                creationTransaction: '0xabc'
            }});
        const wrapper = helper.mountFn(Contract, {
            propsData: {
                hash: '0x123'
            },
            stubs: stubs
        });

        setTimeout(() => {
            expect(wrapper.html()).toMatchSnapshot();
            done();
        }, 2000);
    });

    it('Should display token info if available', async () => {
        jest.spyOn(helper.mocks.server, 'getContract')
            .mockResolvedValueOnce({ data: {
                name: 'ERC20 Contract',
                patterns: ['erc20'],
                tokenName: 'ERC20 Token',
                tokenSymbol: 'ERC',
                tokenDecimals: 18,
                address: '0x123',
                creationTransaction: '0xabc'
            }});

        const wrapper = helper.mountFn(Contract, {
            propsData: {
                hash: '0x123'
            },
            stubs: stubs
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should not display delete option if not admin', async () => {
        jest.spyOn(helper.mocks.server, 'getContract')
            .mockResolvedValueOnce({ data: {
                name: 'ERC721 Contract',
                patterns: ['erc721'],
                tokenName: 'ERC721 Token',
                tokenSymbol: 'ERC',
                tokenDecimals: 18,
                address: '0x123',
                creationTransaction: '0xabc'
            }});

        const wrapper = helper.mountFn(Contract, {
            propsData: {
                hash: '0x123'
            },
            stubs: stubs,
            getters: {
                isUserAdmin: jest.fn(() => false)
            }
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
