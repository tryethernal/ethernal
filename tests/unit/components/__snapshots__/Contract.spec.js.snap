// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract.vue Should display a warning if not a contract 1`] = `
<div class="container container--fluid" hash="0x123">
  <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense v-alert--text warning--text">
    <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-exclamation theme--dark warning--text"></i>
      <div class="v-alert__content">There doesn't seem to be a contract at this address</div>
    </div>
  </div>
  <div class="row mb-1">
    <div class="col-lg-5 col-12">
      <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
        <!---->
        <!---->
        <div class="v-card__subtitle justify-space-between" style="display: flex;">
          <div class="spacer"></div>
          <remove-contract-confirmation-modal-stub></remove-contract-confirmation-modal-stub> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--x-small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-delete theme--light red--text"></i></span></button>
        </div>
        <div class="v-card__text">
          <div class="row">
            <!---->
            <!---->
            <div class="col col-6"><small>Address</small><br> <span class="ml-2"><hash-link-stub type="address"></hash-link-stub></span></div>
            <div class="col col-6"><small>Contract Creation</small><br> <span class="ml-2">N/A</span></div>
          </div>
        </div>
      </div>
    </div>
    <div class="col-lg-7 col-12">
      <div class="row">
        <div class="col col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div class="v-card__subtitle">Balance</div>
            <div align="center" class="v-card__text text-h4">
              0.00000000000001 Ether
            </div>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
            <div class="v-card__subtitle">Transactions</div>
            <div align="center" class="v-card__text text-h4">
              <!---->
              2
              <!---->
            </div>
          </div>
        </div>
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
            <div class="v-card__subtitle">ERC-20 Transfers</div>
            <div align="center" class="v-card__text text-h4">
              3<i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-up-thin theme--light"></i>
              <hr role="separator" aria-orientation="vertical" class="mx-4 v-divider v-divider--vertical theme--light"> 4<i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-down-thin theme--light"></i>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-tabs theme--light">
    <div role="tablist" class="v-item-group theme--light v-slide-group v-tabs-bar v-tabs-bar--is-mobile primary--text">
      <div class="v-slide-group__prev v-slide-group__prev--disabled">
        <transition-stub name="fade-transition"></transition-stub>
      </div>
      <div class="v-slide-group__wrapper">
        <div class="v-slide-group__content v-tabs-bar__content">
          <div class="v-tabs-slider-wrapper" style="transition: none;">
            <div class="v-tabs-slider"></div>
          </div><a tabindex="0" href="#transactions" aria-selected="true" role="tab" class="v-tab v-tab--active" id="transactionsTab">Transactions</a> <a tabindex="0" href="#interactions" aria-selected="false" role="tab" class="v-tab" id="interactionsTab">Read / Write</a> <a tabindex="0" href="#logs" aria-selected="false" role="tab" class="v-tab" id="logsTab">Logs</a> <a tabindex="0" href="#code" aria-selected="false" role="tab" class="v-tab" id="codeTab">Code</a> <a tabindex="0" href="#storage" aria-selected="false" role="tab" class="v-tab" id="storageTab">Storage</a>
        </div>
      </div>
      <div class="v-slide-group__next v-slide-group__next--disabled">
        <transition-stub name="fade-transition"></transition-stub>
      </div>
    </div>
  </div>
  <div class="v-window v-item-group theme--light v-tabs-items">
    <div class="v-window__container">
      <transition-stub name="">
        <div class="v-window-item v-window-item--active" id="transactions">
          <address-transactions-list-stub></address-transactions-list-stub>
        </div>
      </transition-stub>
      <transition-stub name="">
        <!---->
      </transition-stub>
      <transition-stub name="">
        <!---->
      </transition-stub>
      <transition-stub name="">
        <!---->
      </transition-stub>
      <transition-stub name="">
        <!---->
      </transition-stub>
    </div>
  </div>
</div>
`;

exports[`Contract.vue Should display only address & creation if not a token 1`] = `
<div class="container container--fluid" hash="0x123">
  <!---->
  <div class="row mb-1">
    <div class="col-lg-5 col-12">
      <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
        <!---->
        <div class="v-card__title">
          Contract
          <div class="spacer"></div>
          <remove-contract-confirmation-modal-stub></remove-contract-confirmation-modal-stub> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--x-small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-delete theme--light red--text"></i></span></button>
        </div>
        <div class="v-card__subtitle justify-space-between" style="display: flex;">
          <!---->
        </div>
        <div class="v-card__text">
          <div class="row">
            <!---->
            <!---->
            <div class="col col-6"><small>Address</small><br> <span class="ml-2"><hash-link-stub type="address"></hash-link-stub></span></div>
            <div class="col col-6"><small>Contract Creation</small><br> <span class="ml-2">N/A</span></div>
          </div>
        </div>
      </div>
    </div>
    <div class="col-lg-7 col-12">
      <div class="row">
        <div class="col col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div class="v-card__subtitle">Balance</div>
            <div align="center" class="v-card__text text-h4">
              0.00000000000001 Ether
            </div>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
            <div class="v-card__subtitle">Transactions</div>
            <div align="center" class="v-card__text text-h4">
              <!---->
              2
              <!---->
            </div>
          </div>
        </div>
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
            <div class="v-card__subtitle">ERC-20 Transfers</div>
            <div align="center" class="v-card__text text-h4">
              3<i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-up-thin theme--light"></i>
              <hr role="separator" aria-orientation="vertical" class="mx-4 v-divider v-divider--vertical theme--light"> 4<i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-down-thin theme--light"></i>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-tabs theme--light">
    <div role="tablist" class="v-item-group theme--light v-slide-group v-tabs-bar v-tabs-bar--is-mobile primary--text" data-booted="true">
      <div class="v-slide-group__prev v-slide-group__prev--disabled">
        <transition-stub name="fade-transition"></transition-stub>
      </div>
      <div class="v-slide-group__wrapper">
        <div class="v-slide-group__content v-tabs-bar__content">
          <div class="v-tabs-slider-wrapper" style="height: 2px; left: 0px; width: 0px;">
            <div class="v-tabs-slider"></div>
          </div><a tabindex="0" href="#transactions" aria-selected="true" role="tab" class="v-tab v-tab--active" id="transactionsTab">Transactions</a> <a tabindex="0" href="#interactions" aria-selected="false" role="tab" class="v-tab" id="interactionsTab">Read / Write</a> <a tabindex="0" href="#logs" aria-selected="false" role="tab" class="v-tab" id="logsTab">Logs</a> <a tabindex="0" href="#code" aria-selected="false" role="tab" class="v-tab" id="codeTab">Code</a> <a tabindex="0" href="#storage" aria-selected="false" role="tab" class="v-tab" id="storageTab">Storage</a>
        </div>
      </div>
      <div class="v-slide-group__next v-slide-group__next--disabled">
        <transition-stub name="fade-transition"></transition-stub>
      </div>
    </div>
  </div>
  <div class="v-window v-item-group theme--light v-tabs-items">
    <div class="v-window__container">
      <transition-stub name="v-window-x-transition">
        <div class="v-window-item v-window-item--active" id="transactions">
          <address-transactions-list-stub></address-transactions-list-stub>
        </div>
      </transition-stub>
      <transition-stub name="v-window-x-transition">
        <!---->
      </transition-stub>
      <transition-stub name="v-window-x-transition">
        <!---->
      </transition-stub>
      <transition-stub name="v-window-x-transition">
        <!---->
      </transition-stub>
      <transition-stub name="v-window-x-transition">
        <!---->
      </transition-stub>
    </div>
  </div>
</div>
`;

exports[`Contract.vue Should display token info if available 1`] = `
<div class="container container--fluid" hash="0x123">
  <!---->
  <div class="row mb-1">
    <div class="col-lg-5 col-12">
      <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
        <!---->
        <div class="v-card__title">
          ERC20 Contract
          <div class="spacer"></div>
          <remove-contract-confirmation-modal-stub></remove-contract-confirmation-modal-stub> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--x-small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-delete theme--light red--text"></i></span></button>
        </div>
        <div class="v-card__subtitle justify-space-between" style="display: flex;"><span class="success mr-2 v-chip v-chip--no-color theme--light v-size--x-small"><span class="v-chip__content">
                        ERC 20
                    </span></span>
          <!---->
        </div>
        <div class="v-card__text">
          <div class="row">
            <div class="col col-6"><small>Token Name</small><br> <span class="ml-2"><hash-link-stub withtokenname="true" withname="true" hash="0x123" type="token"></hash-link-stub></span></div>
            <div class="col col-6"><small>Token Symbol</small><br> <span class="text-h6 ml-2">ERC</span></div>
            <div class="col col-6"><small>Address</small><br> <span class="ml-2"><hash-link-stub type="address"></hash-link-stub></span></div>
            <div class="col col-6"><small>Contract Creation</small><br> <span class="ml-2">N/A</span></div>
          </div>
        </div>
      </div>
    </div>
    <div class="col-lg-7 col-12">
      <div class="row">
        <div class="col col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div class="v-card__subtitle">Balance</div>
            <div align="center" class="v-card__text text-h4">
              0.00000000000001 Ether
            </div>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
            <div class="v-card__subtitle">Transactions</div>
            <div align="center" class="v-card__text text-h4">
              <!---->
              2
              <!---->
            </div>
          </div>
        </div>
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
            <div class="v-card__subtitle">ERC-20 Transfers</div>
            <div align="center" class="v-card__text text-h4">
              3<i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-up-thin theme--light"></i>
              <hr role="separator" aria-orientation="vertical" class="mx-4 v-divider v-divider--vertical theme--light"> 4<i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-down-thin theme--light"></i>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-tabs theme--light">
    <div role="tablist" class="v-item-group theme--light v-slide-group v-tabs-bar v-tabs-bar--is-mobile primary--text">
      <div class="v-slide-group__prev v-slide-group__prev--disabled">
        <transition-stub name="fade-transition"></transition-stub>
      </div>
      <div class="v-slide-group__wrapper">
        <div class="v-slide-group__content v-tabs-bar__content">
          <div class="v-tabs-slider-wrapper" style="transition: none;">
            <div class="v-tabs-slider"></div>
          </div><a tabindex="0" href="#transactions" aria-selected="true" role="tab" class="v-tab v-tab--active" id="transactionsTab">Transactions</a> <a tabindex="0" href="#interactions" aria-selected="false" role="tab" class="v-tab" id="interactionsTab">Read / Write</a> <a tabindex="0" href="#logs" aria-selected="false" role="tab" class="v-tab" id="logsTab">Logs</a> <a tabindex="0" href="#code" aria-selected="false" role="tab" class="v-tab" id="codeTab">Code</a> <a tabindex="0" href="#storage" aria-selected="false" role="tab" class="v-tab" id="storageTab">Storage</a>
        </div>
      </div>
      <div class="v-slide-group__next v-slide-group__next--disabled">
        <transition-stub name="fade-transition"></transition-stub>
      </div>
    </div>
  </div>
  <div class="v-window v-item-group theme--light v-tabs-items">
    <div class="v-window__container">
      <transition-stub name="">
        <div class="v-window-item v-window-item--active" id="transactions">
          <address-transactions-list-stub></address-transactions-list-stub>
        </div>
      </transition-stub>
      <transition-stub name="">
        <!---->
      </transition-stub>
      <transition-stub name="">
        <!---->
      </transition-stub>
      <transition-stub name="">
        <!---->
      </transition-stub>
      <transition-stub name="">
        <!---->
      </transition-stub>
    </div>
  </div>
</div>
`;

exports[`Contract.vue Should not display delete option if not admin 1`] = `
<div class="container container--fluid" hash="0x123">
  <!---->
  <div class="row mb-1">
    <div class="col-lg-5 col-12">
      <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
        <!---->
        <div class="v-card__title">
          ERC721 Contract
          <!---->
        </div>
        <div class="v-card__subtitle justify-space-between" style="display: flex;"><span class="success mr-2 v-chip v-chip--no-color theme--light v-size--x-small"><span class="v-chip__content">
                        ERC 721
                    </span></span>
          <!---->
        </div>
        <div class="v-card__text">
          <div class="row">
            <div class="col col-6"><small>Token Name</small><br> <span class="ml-2"><hash-link-stub withtokenname="true" withname="true" hash="0x123" type="nft"></hash-link-stub></span></div>
            <div class="col col-6"><small>Token Symbol</small><br> <span class="text-h6 ml-2">ERC</span></div>
            <div class="col col-6"><small>Address</small><br> <span class="ml-2"><hash-link-stub type="address"></hash-link-stub></span></div>
            <div class="col col-6"><small>Contract Creation</small><br> <span class="ml-2">N/A</span></div>
          </div>
        </div>
      </div>
    </div>
    <div class="col-lg-7 col-12">
      <div class="row">
        <div class="col col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div class="v-card__subtitle">Balance</div>
            <div align="center" class="v-card__text text-h4">
              0.00000000000001 Ether
            </div>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
            <div class="v-card__subtitle">Transactions</div>
            <div align="center" class="v-card__text text-h4">
              <!---->
              2
              <!---->
            </div>
          </div>
        </div>
        <div class="col-lg-6 col-12">
          <div class="v-card v-sheet v-sheet--outlined theme--light" style="height: 100%;">
            <div class="v-card__subtitle">ERC-20 Transfers</div>
            <div align="center" class="v-card__text text-h4">
              3<i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-up-thin theme--light"></i>
              <hr role="separator" aria-orientation="vertical" class="mx-4 v-divider v-divider--vertical theme--light"> 4<i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-down-thin theme--light"></i>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-tabs theme--light">
    <div role="tablist" class="v-item-group theme--light v-slide-group v-tabs-bar v-tabs-bar--is-mobile primary--text" data-booted="true">
      <div class="v-slide-group__prev v-slide-group__prev--disabled">
        <transition-stub name="fade-transition"></transition-stub>
      </div>
      <div class="v-slide-group__wrapper">
        <div class="v-slide-group__content v-tabs-bar__content">
          <div class="v-tabs-slider-wrapper" style="transition: none;">
            <div class="v-tabs-slider"></div>
          </div><a tabindex="0" href="#transactions" aria-selected="true" role="tab" class="v-tab v-tab--active" id="transactionsTab">Transactions</a> <a tabindex="0" href="#interactions" aria-selected="false" role="tab" class="v-tab" id="interactionsTab">Read / Write</a> <a tabindex="0" href="#logs" aria-selected="false" role="tab" class="v-tab" id="logsTab">Logs</a> <a tabindex="0" href="#code" aria-selected="false" role="tab" class="v-tab" id="codeTab">Code</a> <a tabindex="0" href="#storage" aria-selected="false" role="tab" class="v-tab" id="storageTab">Storage</a>
        </div>
      </div>
      <div class="v-slide-group__next v-slide-group__next--disabled">
        <transition-stub name="fade-transition"></transition-stub>
      </div>
    </div>
  </div>
  <div class="v-window v-item-group theme--light v-tabs-items">
    <div class="v-window__container">
      <transition-stub name="v-window-x-transition">
        <div class="v-window-item v-window-item--active" id="transactions">
          <address-transactions-list-stub></address-transactions-list-stub>
        </div>
      </transition-stub>
      <transition-stub name="v-window-x-transition">
        <!---->
      </transition-stub>
      <transition-stub name="v-window-x-transition">
        <!---->
      </transition-stub>
      <transition-stub name="v-window-x-transition">
        <!---->
      </transition-stub>
      <transition-stub name="v-window-x-transition">
        <!---->
      </transition-stub>
    </div>
  </div>
</div>
`;
